### ISSUE TITLE
PixiJS Next (v6) Proposal

### ISSUE THREAD

**Issue by bigtimebuddy** [pixijs/pixi.js#6595]
## Overview

@GoodBoyDigital and I had a chat this past weekend about what the next major release of PixiJS would look like. Our overall feeling was that there are no substantial technology changes we want to utilize yet. On the whole, WebGL 2 hasn't yielded as much fruit for us and likely we are still too early for [WebGPU](https://en.wikipedia.org/wiki/WebGPU). Experiments with [wasm](https://en.wikipedia.org/wiki/WebAssembly) and WebWorkers also haven't been huge performance wins. So where does that leave us? Fortunately, for a project that's over seven years old, there's a lot of cruft and an opportunity to start to reduce architectural and API duplication.

## 👯 Double Trouble

There are a handful of areas in PixiJS that are problematic because there are two or more paths for doing essentially the same thing. This bloats the code, exposes us to more bugs, and creates developer confusion. Many of these changes would require picking a winner which likely would mean breaking changes. A major release seems like a good opportunity to make these improvements.

#### Loader cache vs TextureCache/BaseTextureCache/Spritesheet

Whenever you load a resource (image, atlas, etc), there are two caches that are maintained: the Loader resources and the TextureCache. In order to properly remove resources, you need to remove references in both places. In generally, managing resources should happen in one place.

#### Renderer systems vs Renderer plugins

Renderer systems include all the low-level components needed for our rendering pipeline. These include things like framebuffer, mask and filter systems. This is something introduced in v5. Renderer "plugins" however have long been part of PixiJS, these include things like InteractionManager, extract and prepare plugins. Basically, there are two plugin-like systems for Renderer, and we should probably settle on one.

#### BitmapText style vs TextStyle

TextStyle was introduced in v4, however BitmapText still uses the slightly older way of defining style. It would be nice to be able to share the same process for defining text styles for Text or BitmapText or other future text rendering.

#### EventEmitter vs Minisignals vs Runner

Woof, this one is a big one. There are THREE ways of dispatching events in PixiJS. These, like many of the things above have been introduced as various points in the past years, but were never adopted wholesale. Runner is the internal work-horse and was added to v5 and highly optimized for per-frame calls. It's lightweight, but lacks some of the ergonomics and features of signals and events. Currently signals are used by Loader and most other places in Pixi use EventEmitter3. Either way, we have two large dependencies and we should settle on an approach.

#### Abstracting DOM vs Exposing DOM

There are a number of places where the tension between "hey let's just expose how the DOM does it" and "we should create and manage our own abstraction". You can see this tension most notably with InteractionManager, which is trying to be everything for everyone. We conveniently normalize to pointer events (even if the browser lacks them), but also expose all the mouse and touch events. There have been tons of requests for things like bubbling and other DOM-like features. InteractionManager a complex API and hard to build on top of or maintain. Another smaller example if how Text defines things like stroke or gradients that are pass-thru things for context2d rendering of fillText.

#### Global settings vs static class members

`PIXI.settings` was introduced as a way to untangle constants, like `BLEND_MODES` from settable globals like default `RENDER_OPTIONS` or `FILTER_RESOLUTION`. However, `settings` has quickly become the junk-drawer of miscellaneous global or environmental settings. Conversely, there are quite a few places, like Loader, Renderer and Application plugin registration that are set statically. It might be worth considering breaking up `settings` and adding defaults closer to the APIs that use them. For example, `PIXI.settings.RENDER_OPTIONS` becomes `PIXI.Renderer.defaultOptions`.

#### Static construction vs constructors

There are two patterns used all over Pixi for creating objects. `Sprite.from`, `Texture.from`, `RenderTexture.create`. There are also documentation instance of using `new Sprite`, `new Texture` etc. Static construction can be a little less verbose, but it's not clear through documentation or through design what the advantages and when to use one or the other.

#### Global Resolution vs Local Resolution

Resolution, in general, is complex. There have been quite a few patches to how Pixi deals with resolution. It can be set globally, `PIXI.settings.RESOLUTION` or locally, like `Text.prototype.resolution` or `Renderer.prototype.resolution`. Either way, this whole system needs a rethink to try to harmonize how users define and implement various resolutions.

## Summary

We would love to hear other ideas you have which might help to simplify and reduce complexity or dependencies within PixiJS. Please share them in the comments below.

**Comment by andrewstart** [comment]
If we are looking at changing both the loader and the renderer plugins, I would consider tying in the Preload plugin to the loading process (as a load time option), so that it is easier to make uploading heavy assets to the GPU part of loading, preventing the possibility of trying to preload a texture at the same time as "my texture finished loading, I'll toss a Sprite for it on the stage". Right now the Preload plugin can, in theory, do a decent amount to reduce excessively long frames when loading assets, but is kind of hidden away and doesn't get much love.

**Comment by ivanpopelyshev** [comment]
** Documentation **

Yes, it is the main issue. We built an awesome library and plugins, but we dont have enough tutorials.

**Plugins & Modules**

The milestone I see is convertion of all plugins to ES6 and introducing pixi-customize as a main 
bundler for plugins.

If user wants pixi with plugins - he has to create a bundle, and we have to make huge article on how to do that.

** Stage **

We almost separated `core` from `display` and that means we can start coding advanced stage tree, that will support ADDED_TO_STAGE, REMOVED_FROM_STAGE and many other features, like I wanted in #4586

Why not just add those to our main tree? Sorry guys, but I see that our team doesn't have capacity to review all those. In v5 we made only two small changes: 1. `zIndex` inside container 2. `getLocalBounds()` fix.

Even with my huge experience (3 years coding flash tree) - i just cant decide which features should go to vanilla and which are advanced. With each feature in stage tree we'll make code harder and lose people who understand how it works.

Thats why I propose to focus on making custom stage trees! Then we can port best stuff to vanilla. 

** Ports of other renderers **

One of ideas related to custom stage is to port whole ThreeJS on pixijs core :) However we need shader preprocessor system for it.

**Comment by GoodBoyDigital** [comment]
Thanks for writing this all up @bigtimebuddy ! very concise :D

**Loader cache vs TextureCache/BaseTextureCache/Spritesheet**
I think behind the scenes we should have some kind of single Resource Manager. Everything can then ask it for assets and it will be responsible for loading them. That way if we load something once and another part of the code requests the same thing we can return the same item!

**Renderer systems vs Renderer plugins**
I like the idea of everything being a 'system'. Removing the idea of global plugins I think will be good idea too. Make it easy for us (and others) to build custom renderers to suit their needs.
We can still keep our full fat solution but also off a minimal renderer with just core webGL bits!

```
renderer = new Renderer() // just the core

// add some systems bits:
renderer.addSystem(AccessabilitySystem, {debug:true})
renderer.addSystem(InteractionSystem, {refreshRate:1})
renderer.addSystem(PreloadSystem)
```

**BitmapText style vs TextStyle**
Yeah! if we can unify this, great! (especially with what @SukantPal is working on)

**EventEmitter vs Minisignals vs Runner**
See ya later EventEmitter, the 2010's called they want their code back :D

**Abstracting DOM vs Exposing DOM**
Definitely, with this stuff boxed away we open the door to server side rendering! Win win win

**Global settings vs static class members**
👍 

**Static construction vs constructors**
I have always seen `Sprite.from` and `Texture.from` as where the smart logic for determining what type of Resource should be used. 
This also gives us the opportunity to return the same texture from the cache behind the scenes which I love :D
We do the same with `Programs.from` letting us pass through a cache before we generate a new program.
We could rename the RenderTexture to `from` maybe?

using `new` is for people who know what they are doing and want to get a layer deeper and do things like textures with programmatic frames etc.

**Global Resolution vs Local Resolution**
agreed! don't have a opinion on how to best tackle just yet :P

**Comment by SukantPal** [comment]
Following up on @mat - I think there should also be a clean way to “remove” a system. For example, if someone wants to replace the built-in filter system and roll out their own advanced filter pipeline (while having a compatible API).

**Comment by eXponenta** [comment]
I hate static constructors (fabrics):
1. This is explicit constructor, because in all case it should call native constructor.
2. We cannot normally inherit it (static methods not inherit in JS)
3. Users try pass it anything that, like a `div` to texture, for creating textures from it.

I think should be separated fabrics if it needs:
1. TextureUtils.from(...)

**Comment by GoodBoyDigital** [comment]
We may have to agree to disagree on that first one @eXponenta :P
My thinking comes from the idea of the thought process of developers and where things are being intuitive. If they want a texture, they would think of `Texture.from(...` or a Sprite `Sprite.from(` etc. I think so far it has worked incredibly well :)

I agree to agree with your second point mind :D
I like the idea of groups / priority of systems. Are you thinking a system could contain child systems? so we can nest them accordingly?

**Comment by GoodBoyDigital** [comment]
> Following up on @mat - I think there should also be a clean way to “remove” a system. For example, if someone wants to replace the built-in filter system and roll out their own advanced filter pipeline (while having a compatible API).

agree! ideally, you should never need to remove a system when a developer creates a renderer - but the flexibility should be such that you can build a render up from scratch iyswim?

**Comment by SukantPal** [comment]
I don’t think making the built-in systems “nested” would make sense. For user-made ones, it can. I mean renderer.render.projection.setMatrix() would break a lot of stuff.

Yes, you could create a static map of system dependencies (e.g. batch depends on geometry). Then systems can be loaded in order of dependency. I don’t know how system “priority” makes sense here.

Also, I think plugins are a AbstractRenderer type of thing while systems are a webgl Renderer type of thing. If you remove plugins, you’ll need to add system support to AbstractRenderer.

**Comment by GoodBoyDigital** [comment]
all good points to debate! I believe abstract renderer needs to go too :)
we should be able to break out the functionality into systems. I have started to do this
ended up with a 'view' system that did most of what abstract renderer does. If we go full Systems, extending an abstract renderer make little sense! Canvas and WebGL could share those systems

**Comment by SukantPal** [comment]
@GoodBoyDigital that is something I agree on now!

**Comment by qtiki** [comment]
I would like to see proper Promise support for the Loader as well as various other asynchronous methods like `Texture.from(...)` where the Texture is not instantly valid.

**Comment by GoodBoyDigital** [comment]
> I would like to see proper Promise support for the Loader as well as various other asynchronous methods like `Texture.from(...)` where the Texture is not instantly valid.

hmmm. I LOVE promises

**Comment by eXponenta** [comment]
> We may have to agree to disagree on that first one @eXponenta :P

How many times RAW Sprites is used? Extending is more usifull. How u should create extended class for it? U can't d it.
Clear example: `MeshMaterial`. What returned `from` for it? Lol, `Shader`! Why? Becase `from` not overrided and constructor not called.
How i should extend it now for clear implementation? I don't know.

Then.
More examples:

`class Node extends Sprite { ...};`
What should be for Node.from() ? Right, Sprite. Why?.

ok, for resource classes, like as Texture, BaseTexture `from` is more useful, because there are not reason extending for it, it expose Resource API for building textures from anythings.

**Comment by eXponenta** [comment]
> > I would like to see proper Promise support for the Loader as well as various other asynchronous methods like `Texture.from(...)` where the Texture is not instantly valid.
> 
> hmmm. I LOVE promises
`Texture.from` should not  promise, becuse it is static constructor. It should be lazy loaded how it now but waitable if u need it.
u can wait `IResource.load` if u require it.

If u like promises, u should know that promises replace for  callback API.  `from` is not implement callback api.
Then u should know, that IResource already is `thenable` , beacause `load` method is Promise.
We can export link to it inside texture for faster access, and this will not change behaviour of `from`:

```
await texture.load()
```

**Comment by Article19** [comment]
As a developer who came from Director -> Flash - > CreateJS, PIXI was the natural next step, but I have found the documentation at times very difficult to work with. The advice given to understand > intermediate level concepts is to look at the PIXI source code to see what it's doing. While I can appreciate advanced users wanting a peek under the hood, this should not be how average users should be expected to engage with the platform, imo. Better docs, sample code, indications of best-practice, targeted tutorials with the latest PIXI version...all would be a very welcome addition in v6.

By way of feature request: given the limitations of rendering Graphics in WebGL (i.e., non AA edges) it would be interesting to be able to specify whether canvas could be used in the background. We often create complex shapes that get reconverted to images/textures - and being able to benefit from the AA capabilities of canvas (without using the legacy build) would be terrific.

**Comment by eXponenta** [comment]
CanvasRender is completely removed from PIXI standart package.

**Comment by Article19** [comment]
> > We often create complex shapes that get reconverted to images/textures - and being able to benefit from the AA capabilities of canvas (without using the legacy build) would be terrific.
> 
> CanvasRender is completely removed from PIXI standart package, and never be a restored.
> 
> PIXI Legacy is standard PIXI, but include CanvasRenderer, that increases bundle size.
> U never should use Canvas2D in background beacause upload to GPU is very slow. (Text issues)

Understood - thanks very much! We do have a work-around for aliased edges in Graphics that it working. Good luck with v6!

**Comment by eXponenta** [comment]
> Understood - thanks very much! We do have a work-around for aliased edges in Graphics that it working. Good luck with v6!
I use `standard derivatives` for it in my renderer.

**Comment by Article19** [comment]
> > Understood - thanks very much! We do have a work-around for aliased edges in Graphics that it working. Good luck with v6!
> 
> I use `standard derivatives` for it in my renderer.

we generateTexture at twice the resolution we need, then scale it down later. Happy to discuss offline...

**Comment by qtiki** [comment]
> If u like promises, u should know that promises is replace for callback API. `from` is not implement callback api.
> Then u should know, that IResource already is `thenable` , beacause `load` method is Promise.
> We can export link to it inside texture for faster access, and this will not change behaviour of `from`:
> 
> ```
> await texture.load()
> ```

My main problem with the async loading is the Loader itself: https://github.com/englercj/resource-loader

Currently you do this:
```js
loader.load((loader, resources) => {
    // ...
});
```
I would like the load to use a Promise instead:
```js
loader.load().then(({ loader, resources }) => {
    // ...
});
```
Because this is what I _really_ want to write:
```js
const { resources } = await loader.load();
// ...
``` 

It's nice if there are already Promises used internally in Pixi but as I've understood it the resource loader is the preferred method of loading resources. There have been some cases where I've been unsure if a Texture I've created is usable or not and here TypeScript types and Promises would help a lot to understand when I'm actually performing an asynchronous operation.

I'm not that familiar with the insides of Pixi so I don't know if there are other places where Promises could be useful but in general I'd like to see all async code move to use them.

**Comment by ivanpopelyshev** [comment]
@Article19 Here's my take on canvas2d/webgl AA vector graphics problem: https://github.com/gameofbombs/pixi-blit/ , sorry it has only 1 stupid example, but its very powerful :)

CanvasRenderer will never be removed, its just the package is optional. Support of canvas2d features in v5 is much better than in v4 :)

**Comment by Article19** [comment]
> @Article19 Here's my take on canvas2d/webgl AA vector graphics problem: https://github.com/gameofbombs/pixi-blit/ , sorry it has only 1 stupid example, but its very powerful :)
> 
> CanvasRenderer will never be removed, its just the package is optional. Support of canvas2d features in v5 is much better than in v4 :)

Thanks Ivan. Not sure I quite know what do to with your blit solution - but it's good to know that the canvas renderer is there if needed.

**Comment by ivanpopelyshev** [comment]
> Not sure I quite know what do to with your blit solution

you put "models" in it (vector models), and make "sprites" out of them. It automatically creates atlases, depending on what you choose for particular shape : canvas2d or webgl. Right now it generates only mip levels (x1 x2 x4) but later i'll make it "snap" automatically if shape is alone and stays on the same place for a number of frames. For Flash its important because sometimes you need MSAA and sometimes canvas2d is better but it has conflation problem, which can be automatically fixed when you upload it to webgl.

**Comment by englercj** [comment]
I also like the idea of using promises for the loader instead of callbacks. I haven't fully released the new version of resource-loader, so let me experiment with what promises would look like.

**Comment by englercj** [comment]
One thing I'd like to see in v6 is a robust GL layer that isn't related to any pixi features specifically. For my needs `@pixi/core` is too pixi specific to really be useful in isolation.

I'd like to see some GL low-level utilities that pixi is built atop of, similar to the gl-core library we used to have. What I'd expect that library to expose:

- Program - shader compilation, program linking, reflection. No caching.
- Texture - create, destroy, and settings for GL textures. No resource loading.
- IndexBuffer - create, destroy, and manage data of a buffer.
- VertexBuffer - buffer management with attribute setup support.
- Geometry - wrapper for a VAO that maps multiple VertexBuffers and an IndexBuffer together.
- Framebuffer = create, destroy, and settings for GL framebuffers and texture attachments.
- UBOs - uniform block and buffer support for fast uniform uploads.

**Comment by SukantPal** [comment]
@englercj Maybe pixi-gl-core can be updated to the latest developments in @pixi/core?

**Comment by ivanpopelyshev** [comment]
Its better if I'm honest about it: this idea is on par with your old "lets flatten the tree like in phaser v3"  idea (wow its a verse!) for early v5. Your ideas are not bad in theory, but it seems to me that you havent dig in our big progress in the area same way you ignored my progress with layers and trees before.

There's no mentions of RenderBuffers - and they are one of problems that were solved in v5.

The good point on resource loading - it can be easily removed from core, and I was ready to do it in last two months, my PR's are blocked by more prioritized issues and team members free time.

OK, even if we do that and make something like https://github.com/regl-project/regl , and pixi systems would use it instead of pure webgl calls - that will obfuscate code. From my point of view, debugging of systems is pure bliss compared to what we had in v4 with undercooked wrappers. Our biggest achievenement of v5 - @GoodBoyDigital genius solution of mid-level api - will be proven by me in practice soon, I have the biggest pixi project coming in production, 3 years of progress in those systems.

The gl lib with all the enums is also not easy to balance - 
1. too small amount of code will look like a regl clone and it solves nothing 
2. too many features and experience cramped in this architecture - hello, [playcanvas](https://github.com/playcanvas/engine/blob/017c629c72a03561e04a0e49a9473eb734af0fcc/src/graphics/device.js) , 3,5k lines of experience with low-level gl.

@SukantPal @englercj you guys can go ahead and experiment, dont be upset when I will block this.

I really miss the ability to make sub-threads here that exists in slack.

=====

So far the most practical ideas I saw in the thread 
1. @Article19 . Its interesting that it was rebuked by @eXponenta who works in the same field, but if you guys think on the next level - this discussion is very important and you both are right.
2. @qtiki @englercj resource-loader. No comments, it should be done

**Comment by SukantPal** [comment]
@ivanpopelyshev I never talked about making a PR in PIXI.js in this thread.

**Comment by mreinstein** [comment]
Data oriented and ECS exploration in pixi.js would be amazing. I don't use class based object oriented programming in my games.

**Comment by SukantPal** [comment]
@mreinstein That sounds interesting. PixiJS does use a "systems" architecture (which is a part of ECS I believe). What **specific** changes do envision?

**Comment by mreinstein** [comment]
@SukantPal something more like this in terms of architecture: https://github.com/mreinstein/ecs

* **E**ntities have 0 or more components
* **C**omponents are pure data and have no behavior
* **S**ystems have behavior but no data

I'm _not_ saying this specific `ecs` module is perfection and is what pixi should be standardized on, but something like this is the "right" direction, in my opinion. The entire ecs module here is about 100 lines of code, and offers an O(1) iterator at run time for querying for entities within systems.

Scene graphs and object hierarchies end up being a mess, where eventually one ends up facing difficult decisions around questions like "where does this particular piece of logic live?" There are certain bits of logic that cross-cut and don't fit cleanly in one place, and code and logic become entangled in various places in the graph.  Not all relationships in a game/simulation are modeled appropriately with parent/child relationships which is why we get into these situations.

being data-oriented also has really nice performance properties.

there's a really good talk given by Blizzard on ecs provides a solid foundation for overwatch, basically the glue that binds everything together: https://www.youtube.com/watch?v=W3aieHjyNvw

**Comment by bigtimebuddy** [comment]
I think @GoodBoyDigital has done some experiments with ECS patterns and can probably share some thoughts on the performance trade-offs.

**Comment by djmisterjon** [comment]
Very happy with Pixijs at its current state and I think that all was well said.
For my part, I would perhaps just like a more adaptable and easy to serialize architecture including plugins.
I think it the only thing that hurt me using pixiJs.
Good jobs guys and thanks for all your hard work!

**Comment by ivanpopelyshev** [comment]
OK, @GoodBoyDigital was ok with me sharing my opinion: 

0. project that helps people with hard stuff they dont have a chance against otherwise
1. different kind of team, and actual backers from business
2. reputation that is tired to all renderers, not only pixi
3. reputation of a flash successor
4. better support of editors that people like - (RMMV, gdevelop)
5. learning process for those who really want to dig everything, and help with finding jobs/workers

**Comment by GoodBoyDigital** [comment]
Hey peeps! thoughts are coming thick and fast, great to see so many good ideas getting suggested! 

@englercj, you suggestion is cool. I would say that we are so close to this (and in some areas surpassed the concept in v5). The mid level api that powers core is flexible enough that you can do anything in WebGL but streamlining the process as much as possible. At the same time we optimise behind the scenes all the state switching / shader / shader / texture binding. Its super cool - and probably v5 most unknown, underrated feature :)

The following should be enough to build _any_ kind of renderer with a lot less code:

ContextSystem
StateSystem + State
ShaderSystem + Shader
TextureSystem + Texture
GeometrySystem + Geometry
FramebufferSystem + Framebuffer

I know this as we have build an entire 3D engine on it :D. But I will say there is not much out there in terms of tutorials on how to use them. I encourage ya to have a play with it for sure! But yes ultimately totally aligned on the fact we should have and offer a core to build any kind of rendering on from experiments to full fat 2/3D engines. For us to get there I think we could do it with better education and exposing a 'core' renderer with only the above to let people make their own things if they wish.

In terms of ECS. I'm a big fan. Goodboys internal game engine (and our 3D engine) is ECS based. Definitely a great way to structure code. Something that would merit some investigation. Pixi is not far of in terms of structure either. we have a transform object, bounds object etc so not a big thing to twist this into something more formal.

Down sides are that I found the memory footprint goes up when things are split into components, pixi is very sensitive to this. We have optimised the rendering code really well to the point that things like bunny benchmarks slow down as we add properties to the class and its objects :P

The other thing is that we would need to add lot of syntactic suger to it as ECS can get quite wordy to work with and we would not want to complicate the API (one of Pixi's most attractive qualities imo). Both those points could be overcome though i'm sure. 

@ivanpopelyshev , nice ideas there too. Honestly speaking I think we already do. a great job on some of the points already :D Which is a great thing!

project that helps people with hard stuff they don't have a chance against otherwise
- i think we do this right now and should have a strong enough core that we can build plugins as required.

different kind of team, and actual backers from business
- definitely. We have some bigger donations on the horizon for pixi. Would be great to push this area too. Lots of businesses out there using pixi! I'm sure some would pay if we asked :D

reputation that is tired to all renderers, not only pixi
- This is true for a lot of engines and tools already. Something we can be proud of :D Not to blow our own horn.. but pixi is talked about as an industry standard which is something we could never have imagined starting out :) No harm in continuing to improve that :)

reputation of a flash successor
- I would say this is party true already. API is pretty close :P We are certainly lacking tools like a 1st class official editor, something I would love to tackle one day (but this sits out side of pixi v6 scope for sure)

better support of editors that people like - (RMMV, gdevelop)
- Yes, I think this is a good idea. We can continue to achieve this by ensuring our core is easy and flexible enough to cover edge cases our users have right?

learning process for those who really want to dig everything, and help with finding jobs/workers
- This i totally agree with. We lack tutorials and blog posts for lots of the insides of pixi. Especially the juicy midlevel API. Something that we can look to get help from members of the community i'm sure! 

All food for thought! Keep the ideas coming peeps!

**Comment by SukantPal** [comment]
@GoodBoyDigital @bigtimebuddy "Better docs coming for PixiJS 6!" should be a good goal. In the next couple of weeks, we might transition to webdoc. This is what it will bring:

* elimination of .ts to .js precompile step
* faster doc generation times (up to 67% faster already)

Here are some improvements to our documentation template :- 

* Collapsable panes for inherited members
* Separating static & instance members
* (maybe): integration with JSFiddle/CodePen into our documentation for live demos
* exporting an API-structure file so plugin documentation can be "linked" to the core documentation

**Comment by mreinstein** [comment]
while I can appreciate that there are a lot of quality of life improvements for developers in this transition, I don't see really much end-user benefit?

**Comment by SukantPal** [comment]
@mreinstein Your free to propose some ideas!

**Comment by mreinstein** [comment]
I would propose that the API docs are fine, and have been for a while. :) 👍 

Docs that have always been lacking are the ones that actually show how to do things with pixi. 
These would address more of the struggle that new/mid-level pixi users are experiencing.

I've seen several "how to get started with pixi" tutorials over the years, but they are always relegated to people outside the core team, so they aren't considered proper docs. They don't get updated and are left behind as new major releases of pixi are born into the world. For example, the sanctioned tutorials list doesn't actually have many tutorials, and most of these haven't been updated since pixi v4: https://www.pixijs.com/tutorials


So, some actionable ideas around this stuff:
* write more tutorials around how to use pixi
*  create something like a "cookbook" where we take the top 20 tasks in pixi that get asked about and write little "recipe" articles for them.
* commit to keeping these things up to date, and incorporate them into the release process. e.g., pixi v6 shouldn't be considered released until it includes up-to-date tutorials and guides.

If pixi were brand new, I would say no big deal, do this in a follow up release. But pixi is super mature now, it's been around for a while, and it's going on it's 6th major public revision. It's time to give docs more attention and love, and commit to keeping them fresh as part of the release process.

**Comment by SukantPal** [comment]
@mreinstein I see where you're going there. The transition to webdoc is because - JSDoc is not maintained anymore, running JSDoc 3.5 crashes on Node 12, upgrading JSDoc itself produces errors. It's more of a chore I'd say.

* HOWEVER, having a in-house documentation system allows us to create the tools necessary for increasing developer accessibility. For example, I believe that any tutorials should link relevant classes to the API documentation. (Take this guide from Android, for example, that has hyperlinks to classes: https://developer.android.com/guide/components/activities/intro-activities?hl=en)

* webdoc can also help us create tutorials. For example, you can write a wiki in Markdown, and webdoc can output a formatted HTML file (with navigations bars & other goodies)

* Maybe you like the API docs, but I disagree here. The API docs are sub-par right now. For example, inherited members take up majority of the space even though they aren't important in the context of the child class.

* I'd also say that a lot of the API docs don't convey the relations b/w different classes. Take this for example: http://pixijs.download/release/docs/PIXI.Renderer.html - the Renderer's description doesn't tell you the **most** important thing about Pixi's architecture - the **system architecture**. Moreover, the most important method - `render` - is at the bottom. If you're a new developer, the API docs are much more disorganized than I would prefer. The descriptions of most of the classes are not sufficient.

* Now, compare Pixi's most important class `Renderer` and compare it to this documentation of Android's `View` https://developer.android.com/reference/kotlin/android/view/View. The description provided for it is well over 3 page heights. That is something our API can learn from.

* Guides & tutorials are something that aren't incorporated into the API docs. That's sad! Take the Android documentation site - it also has a guides tab: https://developer.android.com/guide . And I think our API docs will benefit from having a "tutorials" section (need not be a tab, but a link to a separate part of the site)

**Comment by mreinstein** [comment]
> having a in-house documentation system allows us to create the tools necessary
> for increasing developer accessibility

I'm not arguing about the merits of upgrading toolchains and making things easier. I think that's great! Where I am a little sad is that the emphasis has pretty much always been on re-working the api docs. I've been following pixi since around the v2 -> v3 release, and this is the 2nd or 3rd or 4th time we've redone the api docs.

It's missing the forest for the trees. When people use pijxi (especially new people) the quality of the api reference guide has _never_ been an obstacle. What has been is finding actual tutorials/walkthroughs on how to use the dang thing. :) By all means if you're super stoked to upgrade this toolset again, go for it! 👏  I'm just telling you as an end user, every time I see us spend time on this rather than making actual docs that get me ramped up on pixi, it just feels like misplaced effort.


> Maybe you like the API docs, but I disagree here

I didn't say I _like_ the api docs, I said they are fine (sufficient) for their purpose. :)


> Take the Android documentation site

I think comparisons are good, but I would say compare to a more similar product. Maybe Unity for example, which always ships very usable docs with all of it's releases: https://docs.unity3d.com/Manual/index.html

comparing unity or android to pixi is not really fair; obviously pixi is a much smaller community with nowhere near the same level of resources. but it's telling that both of our preferred examples show lots of non api documentation. building out this documentation is what matters, not the state of collapsible sidebars in the autogenerated api docs.

**Comment by SukantPal** [comment]
@mreinstein I'll agree with that. I'll make sure I get a few tutorials started out for PixiJS with all the free time I have.

But obviously, both of our experiences have been different regarding PixiJS. I've gained most of my knowledge with Pixi from API docs & working on its source code.

**Comment by irongaze** [comment]
Hey all, thought I'd toss my thoughts in the ring as a relative outsider and noob.  I love a lot of the architectural cleanup being proposed, and can see the value, but I wanted to back up @mreinstein on how very, very important better documentation is.

I've rebuilt my company's core product around Pixi over the past year and a half, and even after working with it extensively, I have no idea how most of it works, or if what I've got working is best practice, or how to extend the system past the basics.

Let's take a small piece of the pie as an example - loading resources.  Promises vs callbacks, ECS vs objects... that is not important to me.  It's fun to build, and of course it's nice to get right, but it won't make or break my Pixi experience.  What *is* important to me are questions like:

1. How do I minimize my memory footprint?  What are overall best practices and common gotchas?
2. How do I release the memory for a texture that is no longer needed?
3. Do I use a single loader?  Spin one up every time I need something new?  How do I slowly stream in content in a Pixi standard way?
4. What do I do with a resource loader once it's finished running?  Does it actually hold the textures and act as a library for Pixi, or does Pixi extract the textures and manage them?  If I don't get rid of the loader, does it hold references to the resources it has loaded and prevent garbage collection?
5. What does the Pixi loader *not* handle?  When should I use something else, and what are some general ideas on how that should be done? For example, the loader won't work for fonts.  Every Pixi user needs loaded fonts - how should I load them?

These are actual issues for using Pixi in production, and there is virtually no information on most of these questions in the official docs, the forums, Stack Overflow, blogs... My google-fu is pretty strong, I've done few runs through the source code.  I still have no idea how to manage resource loading and unloading properly.

This is just one example.  I could go on for days about the many mystifying challenges to building something solid with Pixi   Texture.from - ok, it works great in my little sample.  Now I want to use it in production.  Bad idea?  Best practice?  No clue.  I use it for two buttons, and modify the frame on one, and now the other button changes.  Why?  No answers.  I know that somewhere under the hood there is an actual image loaded by the browser.  Somehow the actual pixels get moved to the GPU.  There's a texture, and a base texture.  Why?  From my button failure, I guess there's some caching going on.  How does that work?  Why do I get errors in the console when I load the same image twice?  ARGH.  And this is the *basic* stuff.  

Trying to figure out how to customize the render pipeline has got me climbing the walls.  And I wrote an entire 3D rendering pipeline in C++.  I've been programming for over 35 years, in literally dozens of languages.  The concepts aren't the hard part, figuring out what problems Pixi solves, and what it doesn't, and how to close that gap - THAT's the hard part.

**TL;DR - Pixi the rendering system is fantastic.  Pixi the development ecosystem is not.**  As you contemplate priorities, I'd urge you to consider nurturing the ecosystem as a higher priority than adding features or reorganizing internals.

A few last things: This comes from a place of deep love.  You guys are terrific and I deeply value the tools you've built.  They let me build games to help kids read with nothing but a web browser and some moxie.  It's amazing and I am incredibly grateful for everyone who has contributed.  Please don't take this as criticism, so much as a cry for help from someone who wants to go beyond appreciating what you've built to understanding and mastering it.

**Comment by bigtimebuddy** [comment]
Thank you @irongaze and @mreinstein for your thoughts. It's a very helpful perspective. I've had conversations with various people on the team and in the community about this. I think there's been an overwhelming consensus that we need to do better in terms of usage documentation and best practices. We have bits and pieces but it's not authoritative or as comprehensive as it needs to be.

I see the main challenge here is that PixiJS is a volunteer-driven project and we don't have a ton of say in who decides to contribute. In general, the project has self-selected for people who are primarily interested in solving technical, not communication, challenges.  Basically, we don't have a ton of people that are excited to build the type of documentation people need. It requires a different set of skills, is time-consuming, and sometimes requires a deep knowledge of how things actually work.

The challenge is finding and partnering with the right people who can fill these deficits. We have limited funds on Open Collective, which maybe how we approach this problem. That, or, wait for the right person to show up. Either way **this is a people-problem not a computer-problem**. A lot of what I've proposed above is knowing that we have people who are willing and able to execute on, say, improving some of the architecture issues.

If anyone has any ideas or knows of people who might be willing to help, please reach out.

**Comment by irongaze** [comment]
I'd be happy to work on the problem.  I think the first step would be figuring out where official guides would live, and drawing up a list of topics that need covering.  Then we need some kind of versioned guide publication platform to host the content, and a working group of folks who'd be willing to be technical advisors.  If folks have ideas on any of the above, I'd love to hear about it, but if not I'll start pulling together resources tonight.  Is the slack channel the preferred communication channel for this type of work?

**Comment by SukantPal** [comment]
@irongaze @bigtimebuddy I think JSDoc had a feature for creating "tutorials" that we didn't use. Maybe we could incorporate it into webdoc.

The tutorials themselves would be written in Markdown & stay in the repo. This would ensure a tight relation b/w updates to the code & updates in the tutorials.

As for versioning, I believe we already have versioned docs based on the release. For example v4.7.2 docs are here: https://pixijs.download/v4.7.2/docs/PIXI.Transform.html

**Comment by bigtimebuddy** [comment]
@irongaze Thanks for being willing to help! Let's connect on Slack. 

Also, here's a relevant list that @GoodBoyDigital compiled as a list of potential example for v5. I think this is a good start place: https://github.com/pixijs/pixi.js/projects/10

**Comment by GoodBoyDigital** [comment]
@irongaze awesome :)

**Comment by irongaze** [comment]
@SukantPal Agreed we want to have the docs in Git.  But the versioning I'm talking about would be something like going to guides.pixijs.io, and having at the top of the left nav "Version 4", "Version 5", "Version 6" to let folks view guides targeted to their desired major version.  I'm sure there are folks still on v4, and when v6 goes live, we'll need to update each guide as needed, while preserving the older guides for folks who e.g. need a plugin that hasn't yet been ported.

@bigtimebuddy Will do re: Slack.  I'm an east-coaster who works 9-6pm or so, will check in later this evening.

**Comment by SukantPal** [comment]
@irongaze For different versions, the version menu could link to the site for the different version. 

But I'm not sure how we would update the v4 site in retrospect to add guides for v4. Creating a menu for v5 & v6 guides would be easier (since we can still edit dev-branch site before going to v6).

**Comment by mreinstein** [comment]
I think it will be hard enough improving docs for v6. If we were a larger team I'd suggest we go backfill v4 and v5 guides but it just doesn't make sense with our resources. I'd rather see that effort poured into making v6 docs amazing, and maintaining that commitment going forward.

As an unintentional side effect of this, it creates a really powerful incentive to migrate to the new version. :)

**Comment by irongaze** [comment]
I wasn't suggesting v4 docs, though I think v5 may be warranted unless v6 is going to be faster to market than this thread suggests.  :-)  Just an example of how I think we want to future-proof the guides system.

**Comment by bigtimebuddy** [comment]
For many thing related to docs or examples, we don’t have to wait for v6. We should try to tackle some of that now and setup some mechanisms to make it easier to maintain into the future.

**Comment by bigtimebuddy** [comment]
Just a quick note: I want to get away from calling this "v6". The number is irrelevant... what's outlined above constitutes our next big push, whatever that number is.

**Comment by mreinstein** [comment]
I actually think it might be better to have the actual version number in the issue. Someday, pixi v6 will be released, and the issue will be closed. Being able to look through the issue history and not see every single release issue aliased to `pixi next` has some merit.

Perhaps some combination `PixiJS Next (v6) Proposal` ? I dunno.

**Comment by zbot473** [comment]
Can we get resource scale in the loader?

**Comment by CosmoMyzrailGorynych** [comment]
I join the @irongaze, the loader's API is so obscure in practice that I don't heckin' know how it works, and that's while having a game engine+IDE built on top of pixi.js. There is a huge gap between loading and forgetting use case and dynamic loading/unloading for bigger games.

**Comment by donthebike** [comment]
What I really mis (or didn't find yet if it exists) is a good and performant picking mechanism in pixijs, such as described by https://webglfundamentals.org/webgl/lessons/webgl-picking.html and in https://webglfundamentals.org/webgl/lessons/webgl-picking.html
Would it maybe be possible to add this in a next release?
For the rest congrats on an extremely good open source product!

**Comment by eXponenta** [comment]
> What I really mis (or didn't find yet if it exists) is a good and performant picking mechanism in pixijs, such as described by https://webglfundamentals.org/webgl/lessons/webgl-picking.html and in https://webgl2fundamentals.org/webgl/lessons/webgl-picking.html
> Would it maybe be possible to add this in a next release?
> For the rest congrats on an extremely good open source product!

As ex-member of PIXI team not sure that this is good idea.
Rendering a scene graph more that 1 time per render call is expensive. And this method will uncompobility with external libs that use custom rendering process.
Will problems with mask/filters and phases that use renderTextures.

**Comment by ivanpopelyshev** [comment]
I use that color-picking without extra pass , its just every shader contains `if (gl_FragCoord === u_mouse_pos)` , and it works. Yes, readPixels stops execution for some time, in case mouse moved. There are **many** downsides and corner cases, I can make a plugin about it, but we need to make a ShaderLib support , that im pushing for a long time. With ShaderLib this thing will be possible.

**Comment by donthebike** [comment]
**@eXponenta** 

Tx for link. Quadtrees are great indeed, but for static content. For dynamic content, they have to be recalculated often, right?

**@ivanpopelyshev** 

Thanks for that. Is the complete source for your one-pass solution available somewhere?

Regards

**Comment by CosmoMyzrailGorynych** [comment]
> **Global Resolution vs Local Resolution**

I want to share my experience with pixi.js v4/v5 in building a logical and usable resolution system, and propose solutions to this issue. I'm working on a game engine with a proper IDE — it is a pretty high-order tool that **prioritizes ease of use for end-users** (for game developers).

Before everything, let's make sure that we are on one page. We have three coordinate systems:

* physical resolution as a number of pixels on your physical screen;
* web resolution as CSS' `px` unit;
* designed resolution. For most non-gamey apps, it will be identical to web resolution. For most games, it will be a static rectangle (a viewport size in a game's coordinates).

We also have canvas size (the same as renderer's size). Ideally, it matches either physical resolution or web resolution, depending on performance and requirements for crisp results.

So,

* pixi.js needs a camera entity that manages scene scaling and canvas size. A developer specifies the target canvas size in CSS pixels and pixel density modifiers (1 or window.devicePixelRatio).
* A developer works in *designed resolution*. The region that is shown is determined by a camera's viewport size. Depending on a developer's needs, they either use static viewport size or use the canvas' dimensions in CSS pixels. In any case, developers get crisp results;
* With proper camera system, further `resolution` tweaks are redundant;
* `PIXI.Text`'s resolution parameter is a hacky solution to a distinct problem — the way we _can_ draw text on web canvases. Introducing vector fonts is a solution to the root problem.

I have a camera entity in my engine that makes resolution management for users... well, they actually don't think about the resolution at all. [Here is the module](https://github.com/ct-js/ct-js/blob/develop/app/data/ct.release/camera.js#L367) that manages CSS scaling, renderer's size, and camera properties for different use cases ­— expanding with no scaling, covering the whole screen, scaling with letterboxing.

Scene graph has layers (PIXI.Containers) as direct children to the stage, and this function positions them according to the camera: https://github.com/ct-js/ct-js/blob/develop/app/data/ct.release/camera.js#L367

## PIXI.Text issue

In my view, **`resolution` for text labels is a patch** over the issue that previously there was no way to draw scalable fonts, in the way they are drawn everywhere: as vector shapes. But now, with [opentype.js](https://github.com/opentypejs/opentype.js/), we can solve the root issue, and it doesn't even require an offline builder.

I used opentype.js and HTMLCanvas to create bitmap font texture, and the results are crisp and precise, even for pixel fonts. Sounds absurd considering that plain `CanvasContext2D.fillText` produces blurry results... For truly vector fonts, the results will be even better.
 
## `PIXI.settings.RESOLUTION` and `Renderer.prototype.resolution`

Should be deprecated: each Renderer instance should have its own independent resolution parameter, as one can have several pixi.js apps on a web page that have different requirements for clarity on Hi-DPI screens. Tweaking this parameter will also allow coding a dynamic resolution that adapts for poor graphics performance (similar to dynamic resolution in UE4).

**Comment by cjsjy123** [comment]
Will support 3D UI like this?https://www.fairygui.com/threejs-demo/3d/

**Comment by irongaze** [comment]
Out of this issue, I've started working with the core team to build a set of guides for PixiJS to cover everything from in-depth dives on the basics, to common problems putting PixiJS into production, and ideally if time permits, advanced topics that go beyond typical use cases.  I'd welcome suggestions for areas that need clarification or special treatment!

But as I'm working on the guides, one place that could really use some clarity is the set of functions used in the Graphics class to build and manage the array of primitives in the GraphicsGeometry buffer.  The functions are currently named `drawRect`, `drawCircle`, etc., which is unnecessarily confusing.  I'm working on a PR to improve the API docs, but I'd like to propose a small change for v6: move from "draw" to "add"

Part of what makes Graphics so confusing is our use of "draw" as the verb for the core functions, when in fact no drawing is happening.  Instead, the drawX functions are constructing primitives and storing them for later use - in rendering, yes, but also in masking and hit testing.

I'd propose in v6 that we deprecate the drawX versions of the Geometry API and replace them with e.g. `addRect`, `addCircle` and so forth.  Similarly, I'd propose changing `clear` to be `reset`.  You don't clear an array of primitives, you reset it.  You clear pixels.

I know there's history here, and we're mirroring the Canvas API's naming, but since what we're doing *isn't what Canvas is doing*, I feel that this small change would make a big difference in making the Graphics class less mysterious and more useful.

**Comment by mreinstein** [comment]
>  I've started working with the core team to build a set of guides for PixiJS to cover everything

That's **amazing**, thank you! ❤️ 


>  this small change would make a big difference in making the Graphics class less mysterious and more useful.

I understand what you're saying about `addX` vs `drawX` but I don't think it makes sense to break the pixi API just for the sake of a name. I could just as easily imagine 6 months from now someone making the exact opposite claim that `draw` is more intuitive than `add`. I'm not saying I agree with one form vs the other, it just doesn't ultimately matter, at least not strongly enough to break the API over this. API stability is a _good_ thing.

**Comment by CosmoMyzrailGorynych** [comment]
> API stability is a good thing.

@mreinstein, absence of ambiguity and intuitive API is objectively better.
I vote for `addRect` and similar methods as people coming from various game engines expect imperative shape drawing instead of an object-oriented one. They add shapes or other entities at each frame and then wonder why their game crashes after a minute or so. This is a common issue at ct.js, Pixi.js-based game engine.

I do vote against `clear` -> `reset` change, as I see no lexical issues here. At least from my English-as-a-second-language point of view :D But maybe `empty` will be even a better match.

**Comment by mreinstein** [comment]
> absence of ambiguity and intuitive API is objectively better.

It is unclear to me that `addXXX` is objectively better or worse than `drawXXX`. It's opinion either way.

Maintaining software is a lot of work! I don't want name changes that break the API unless there is real improvement to the end user ("real" defined here as faster code or less code or better memory usage, etc.) All this does is break existing pixi code, over what is basically an opinion about function names. Please do not break stuff for the sake of it.

**Comment by bigtimebuddy** [comment]
In general, methods and properties are easily for us to deprecate (and maintain backward compatibility). These are common things we've done for releases, including feature/minor releases.

I think that @irongaze's point is that draw* methods in Graphics don't reflect the underlying functionality or represent the best mental model for users. It's not like draw methods on context2d which actually rasterizes pixels whereas Graphics maintains a bunch of primitives internally. I'm not opposed to this change, but like I said it trivial to change.

**Comment by mreinstein** [comment]
> methods and properties are easy for us to deprecate (and maintain backward compatibility)

I understand that there is an established mechanism for deprecating things, but my question still stands: why is this a good thing?

Deprecating something in release `X`  usually means developers in versions `X+1` can make that deprecation a breaking change because users have already been warned in `X` release.  

Just to be clear I am not advocating for _never_ making breaking changes to an API. I maintain dozens of modules and have to make breaking changes sometimes.  What I _am_ advocating for is not breaking stuff just for the sake of it. Having API stability is more valuable than changing the name of a function. It reeks of junior project syndrome where things constantly break because people won't stop "screwing with things". 🤷

**Comment by CosmoMyzrailGorynych** [comment]
> Maintaining software is a lot of work!

I know it by myself, ok?

> "real" defined here as faster code or less code or better memory usage, etc

Oh yes, UX improvements that make overall usage less error-prone are definitely surreal, useless improvements. Such methods cost hours of debugging for end users, as well as grey hair.

> It reeks of junior project syndrome where things constantly break because people won't stop "screwing with things".

As **bigtimebuddy** said, it is not a big effort for pixi.js maintainers. It is not a big one for users, either, as they will have a couple of years to migrate their code (if they ever plan to update pixi.js). Designing API filled by quirks known by developers only is a junior thing, not designing API for end users.

**Comment by mreinstein** [comment]
> UX improvements that make overall usage less error-prone are definitely surreal, useless improvements

Breaking software that worked just fine before, for the sake of changing a name is _not_ a user experience improvement, it's a regression.

> Designing API filled by quirks

All of these discussion points make the assumption that one naming scheme is better than another, and I think that's from _your_ perspective. It is not a "universal truth" that either of those names are technically superior. If people are so concerned with making the pixi codebase easier to understand, I think that's wonderful, and that's something that good documentation will improve. Pulling the rug out from under people for the sake of naming fads is just immature product ownership.

**You don't need to add more barriers to pixi, make changes thoughtfully by considering real world tradeoffs.**

**Comment by bigtimebuddy** [comment]
This thread is off topic. Can we not have abstract discussions about project maintenance? We do our very best to not break people when unnecessary and we also rename things all the time. Stability is the most important quality for large projects like ours and we try our best.

The issue is "draw" vs "add" in Graphics method names. My suggestion is someone make a PR to change this and we continue this discussion there about the trade-offs of changing a high level API.

Thank you everyone for your thoughts. If anyone has strong opinions about things they'd like to see in a major release, welcome more feedback.

**Comment by mreinstein** [comment]
> This thread is off topic.

Well, the topic is "what should go into v6" and I think it would be great if v6 avoided unnecessarily breaking stuff. That seems like a pretty fundamental goal for the next release.

> someone make a PR to change this and we continue this discussion there about the trade-offs of changing a high level API.

I think that's totally fair. 👍  I'll refrain from contributing to the particular topic in this thread. thanks for weighing in.

**Comment by Bouh** [comment]
#### BitmapText style vs TextStyle


I would suggest to have a[ Text object visually identical to the BitmapText](https://github.com/pixijs/pixi.js/issues/6884#issue-702266569) object for V6.

Let me explain, Text and BitmapText both render text, but the rendering is different because not all styles are taken into account, and also Matt was telling me that this was because they use different rendering techniques.

An average user shouldn't have to worry about that. 
What it is is two objects to make text, switch from Text to BitmapText and inverse should be logic and possible for the an average user without teak the objects.

Currently objects have different visual despite the same style.

And all styles possible in Text are not yet supported in BitmapText.
The differences between these objects are still large.

The average use currently need to rework their styles if he switch of class text. (PIXI.Text-PIXI.BitmapText)

I would love see an easy way to switch of text tech.

**Comment by themoonrat** [comment]
I just don't see how this is possible, unfortunately. 
Photoshop will never match identically with Text ... because we're just using the Canvas API, so rely on how that renders text, and that's going to be different to Photoshop, as they've different engines.
And Text to BitmapText ... they're fundamentally different things. Normal text has it's own kernings and such that a bitmap version, auto created, just doesn't have.
So maybe we can add support, or an example, of how you can manually supply your own kernings data for each letter, and better support line height and stuff to get them closer.But that can never be an automatic thing; manual work will be required for each font style generated.

So _identical_; I unfortunately don't think that's possible. Closer, yes, with some manual settings provided, but identical... :/

**Comment by IljaDaderko** [comment]
While discussion is ongoing about BitmapText:

I came across similar issue as posted in https://github.com/pixijs/pixi.js/issues/6823 where bitmap text doesn't have support for line heights. The only workaround I came up with was to separate each line of text into its own container and add top offsets.
By default textline height for text squeezes everything very close together.

Is it possible / viable to add this style in v6?

**Comment by st0nerhat** [comment]
**Loader cache vs TextureCache/BaseTextureCache/Spritesheet**
Probably way too late to the part on this, but wanted to +1 having a centralized resource management system with the concept of Resource objects that are reference counted handles to underlying resources. The rationale for this level of sugar is that in larger games (1+ hour playtime, many levels, scenes) you generally want to purge a part of the cache at key load points. However, without some hackery, there isn't a clean way to keep track of what resources can be purged from the cache that won't either generate an error or an immediate disk load, and which ones will. Other major game engines utilize reference counting. You could also implement a default garbage collector like exists for GLTextures that is frame-count or time based and could be overridden with a custom implementation.

Down the road, the Resource Manager could be extended to handle things like manifests for being able to show an accurate loading progress bar for example.
